File: android/app/src/main/java/com/example/subscriptionguardian/HomeWidgetProvider.java
package com.example.subscriptionguardian;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.SharedPreferences;
import android.widget.RemoteViews;
import es.antonborri.home_widget.HomeWidgetPlugin;

public class HomeWidgetProvider extends AppWidgetProvider {
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.home_widget);
            SharedPreferences prefs = context.getSharedPreferences("HomeWidgetPreferences", Context.MODE_PRIVATE);
            String upcomingRenewals = prefs.getString("upcoming_renewals", "No upcoming renewals");
            views.setTextViewText(R.id.widget_content, upcomingRenewals);
            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }
}

File: android/app/src/main/kotlin/com/example/subscriptionguardian/HomeWidgetProvider.kt
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.SharedPreferences
import android.widget.RemoteViews
import com.example.subscriptionguardian.R
import es.antonborri.home_widget.HomeWidgetProvider

class HomeWidgetProvider : AppWidgetProvider() {
    override fun onUpdate(context: Context?, appWidgetManager: AppWidgetManager?, appWidgetIds: IntArray?) {
        if (context == null || appWidgetManager == null || appWidgetIds == null) return
        
        val prefs = context.getSharedPreferences("home_widget", Context.MODE_PRIVATE)
        appWidgetIds.forEach { widgetId ->
            val views = RemoteViews(context.packageName, R.layout.home_widget_layout).apply {
                val upcomingRenewals = prefs.getString("upcoming_renewals", "No upcoming renewals")
                setTextViewText(R.id.widget_data, upcomingRenewals)
            }
            appWidgetManager.updateAppWidget(widgetId, views)
        }
    }
}
